app/assets/javascripts/admin/addon/components/settings-editor.js

```diff
@@ -0,0 +1,188 @@
+import Component from "@ember/component";
+import { action, computed } from "@ember/object";
+import { tracked } from "@glimmer/tracking";
+import { ajax } from "discourse/lib/ajax";
+import { inject as service } from "@ember/service";
+import I18n from "I18n";
+
+export default class SettingsEditor extends Component {
+  @service dialog;
+
+  @tracked editedContent = null;
+  @tracked errors = [];
+  @tracked saving = false;
+
+  didInsertElement() {
+    this.saving = false;
+    this.errors = [];
+  }
+
+  @computed("editedContent", "saving")
+  get saveButtonDisabled() {
+    return !this.documentChanged || this.saving;
+  }
+
+  @computed("editedContent")
+  get documentChanged() {
+    try {
+      if (!this.editedContent) {
+        return false;
+      }
+      const editedContentString = JSON.stringify(
+        JSON.parse(this.editedContent)
+      );
+      const themeSettingsString = JSON.stringify(this.condensedThemeSettings());
+      const same = editedContentString.localeCompare(themeSettingsString);
+      return same !== 0;
+    } catch (e) {
+      return true;
+    }
+  }
+
+  _theme() {
+    return this.model.model;
+  }
+
+  condensedThemeSettings() {
+    return this._theme().settings.map((setting) => ({
+      setting: setting.setting,
+      value: setting.value,
+    }));
+  }
+
+  @computed("theme")
+  get editorContents() {
+    return JSON.stringify(this.condensedThemeSettings(), null, "\t");
+  }
+
+  set editorContents(value) {
+    this.errors = [];
+    this.editedContent = value;
+  }
+
+  validateSettingsKeys(settings) {
+    return settings.reduce((acc, setting) => {
+      if (!acc) {
+        return acc;
+      }
+      if (!("setting" in setting)) {
+        return false;
+      }
+      if (!("value" in setting)) {
+        return false;
+      }
+      if (Object.keys(setting).length > 2) {
+        return false;
+      }
+      return true;
+    }, true);
+  }
+
+  @action
+  async save() {
```

Comment:
It would be useful to have a component test and focus on this section. There's a fair bit of logic here that would be good to have test cases for, it would help prevent regressions in the future. 

I am thinking of cases for: 

- add invalid JSON, hit save, ensure correct error is shown
- remove a setting, hit save, ensure correct error is shown
- add a setting name, hit save, ...

Reply:
done